pennylane/plugins/default_qubit.py

```diff
@@ -309,79 +309,82 @@ class DefaultQubit(QubitDevice):
                                   "on a {} device.".format(operation.name, self.short_name))
-            if input_state.ndim == 1 and n_state_vector == 2**len(wires):
 
-                # generate basis states on subset of qubits via the cartesian product
-                tuples = np.array(list(itertools.product([0, 1], repeat=len(wires))))
+            if isinstance(operation, QubitStateVector):
+                self.apply_state_vector(wires, par)
 
-                # get basis states to alter on full set of qubits
-                unravelled_nums = np.zeros((2 ** len(wires), n), dtype=int)
-                unravelled_nums[:, wires] = tuples
+            elif isinstance(operation, BasisState):
+                self.apply_basis_state(wires, par)
 
-                # get indices for which the state is changed to input state vector elements
-                nums = np.ravel_multi_index(unravelled_nums.T, [2] * n)
-                self._state = np.zeros_like(self._state)
-                self._state[nums] = input_state
             else:
-                raise ValueError("State vector must be of length 2**wires.")
-            self._first_operation = False
-            return
-
-        if isinstance(operation, BasisState):
-            # length of basis state parameter
-            n_basis_state = len(par[0])
-
-            if not set(par[0]).issubset({0, 1}):
-                raise ValueError("BasisState parameter must consist of 0 or 1 integers.")
-            if n_basis_state != len(wires):
-                raise ValueError("BasisState parameter and wires must be of equal length.")
-            if not self._first_operation:
-                raise DeviceError("Operation {} cannot be used after other Operations have already been applied "
-                                  "on a {} device.".format(operation.name, self.short_name))
+                A = self._get_operator_matrix(operation.name, par)
+                self._state = self.mat_vec_product(A, self._state, wires)
+
+        # store the pre-rotated state
+        self.state = self._state
+
+        # apply the circuit rotations
+        for operation in rotations:
+            wires = operation.wires
+            par = operation.parameters
+            A = self._get_operator_matrix(operation.name, par)
+            self._state = self.mat_vec_product(A, self._state, wires)
+
+    def apply_state_vector(self, wires, par):
+        input_state = np.asarray(par[0], dtype=np.complex128)
+
+        if not np.isclose(np.linalg.norm(input_state, 2), 1.0, atol=tolerance):
+            raise ValueError("Sum of amplitudes-squared does not equal one.")
+
+        n_state_vector = input_state.shape[0]
 
+        if input_state.ndim == 1 and n_state_vector == 2**len(wires):
+            # generate basis states on subset of qubits via the cartesian product
+            tuples = np.array(list(itertools.product([0, 1], repeat=len(wires))))
```

Comment:
Having the following renamings could help readability here:
``tuples`` -> ``basis_states``,
``unravelled_nums`` - > ``unravelled_indices``,
``ravelled_nums`` - > ``ravelled_indices``